server/src/main/java/org/opensearch/telemetry/tracing/channels/TraceableTransportChannel.java

```diff
@@ -0,0 +1,100 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.telemetry.tracing.channels;
+
+import org.opensearch.Version;
+import org.opensearch.common.util.FeatureFlags;
+import org.opensearch.core.action.ActionListener;
+import org.opensearch.core.transport.TransportResponse;
+import org.opensearch.telemetry.tracing.Span;
+import org.opensearch.telemetry.tracing.SpanScope;
+import org.opensearch.telemetry.tracing.Tracer;
+import org.opensearch.telemetry.tracing.listener.TraceableActionListener;
+import org.opensearch.transport.TcpChannel;
+import org.opensearch.transport.TransportChannel;
+
+import java.io.IOException;
+
+/**
+ * Tracer wrapped {@link TransportChannel}
+ */
+public class TraceableTransportChannel implements TransportChannel {
+
+    private final TransportChannel delegate;
+    private final Span span;
+    private final Tracer tracer;
+
+    private final TcpChannel tcpChannel;
+
+    private final static ActionListener<Void> DUMMY_ACTION_LISTENER = ActionListener.wrap(() -> {});
+
+    /**
+     * Constructor.
+     * @param delegate delegate
+     * @param span span
+     * @param tracer tracer
+     */
+    public TraceableTransportChannel(TransportChannel delegate, Span span, Tracer tracer, TcpChannel tcpChannel) {
+        this.delegate = delegate;
+        this.span = span;
+        this.tracer = tracer;
+        this.tcpChannel = tcpChannel;
+    }
+
+    /**
+     * Factory method.
+     *
+     * @param delegate delegate
+     * @param span     span
+     * @param tracer tracer
+     * @return transport channel
+     */
+    public static TransportChannel create(TransportChannel delegate, final Span span, final Tracer tracer, final TcpChannel tcpChannel) {
+        if (FeatureFlags.isEnabled(FeatureFlags.TELEMETRY) == true) {
+            tcpChannel.addCloseListener(TraceableActionListener.create(DUMMY_ACTION_LISTENER, span, tracer));
```

Comment:
This dummy listener is wrapped in TraceableActionListener, which will take care of closing the span if remains un-ended. Or I am missing something here?

Reply:
doing multiple ends don't have any side effects. We can do that explicitly also but for that we need to expose the Span::hasEnded method.

Reply:
Ah! I missed that part where sendResponse is called after the channel is closed. In that case you are right we would be propagating the already ended span. We need to put a Span::hasEnded check in the TransportChannel::sendResponse as well.

Reply:
@Gaganjuneja I was fixing one of the edge cases with HttpChannel when is was not closed and it turned out it would be very difficult to enforce that the span is ended only once. The Otel spec is explicit with what happens there (https://opentelemetry.io/docs/specs/otel/trace/api/#end)

> Implementations SHOULD ignore all subsequent calls to End and any other Span methods, i.e. the Span becomes non-recording by being ended (there might be exceptions when Tracer is streaming events and has no mutable state associated with the Span).

> End MUST NOT inactivate the Span in any Context it is active in. It MUST still be possible to use an ended span as parent via a Context it is contained in. Also, any mechanisms for putting the Span into a Context MUST still work after the Span was ended.

So if stick to Otel semantic, we should be fine with both issues - ending span multiple times and propagating the context for already ended span. But it may not play well with other implementations (OpenTelemetry, Zipkin, ...)